crates/engine/tree/src/tree/payload_processor/prewarm.rs

```diff
@@ -134,26 +143,40 @@ where
         // spawn execution tasks.
         self.spawn_all();
 
+        let mut final_block_output = None;
         while let Ok(event) = self.actions_rx.recv() {
             match event {
                 PrewarmTaskEvent::TerminateTransactionExecution => {
                     // stop tx processing
-                    self.pending.clear();
+                    self.ctx.terminate_execution.store(true, Ordering::Relaxed);
                 }
                 PrewarmTaskEvent::Outcome { proof_targets } => {
-                    // completed a transaction, frees up one slot
+                    // completed executing a set of transactions
                     self.send_multi_proof_targets(proof_targets);
+
+                    // decrement the number of tasks left
+                    self.prewarm_tasks_left -= 1;
+
+                    if self.prewarm_tasks_left == 0 && final_block_output.is_some() {
+                        // all tasks are done, and we have the block output, we can exit
+                        break
+                    }
                 }
                 PrewarmTaskEvent::Terminate { block_output } => {
-                    // terminate the task
-                    if let Some(state) = block_output {
-                        self.save_cache(state);
-                    }
+                    final_block_output = Some(block_output);
 
-                    break
+                    if self.prewarm_tasks_left == 0 {
+                        // all tasks are done, we can exit, which will save caches and exit
+                        break
+                    }
```

Comment:
I think this is fine, I'm only concerned about the case where the spawned prewarm task would panic which would then violate this, so

I think, now that we're spawning all tasks upfront, we should drop this one

https://github.com/paradigmxyz/reth/blob/ff04aee216fb7413501244bf7cfc721d27dd35e1/crates/engine/tree/src/tree/payload_processor/prewarm.rs#L47-L48

so that we can't have a case where this loop remains active indefinitely?

Reply:
yes, I think dropping the `actions_tx` after we `spawn_all` would be the easiest thing here that would work

Reply:
added `drop_actions_tx`, to get around partial move issues